File: dashboard/backend/services/epic_flow.py
"""Phase 2 (dashboard-epic-phase-view-p2) — assemble an Epic -> Phase flow graph.

Pure functions over p1's `EpicGroup`: nodes (one per Phase, carrying the current
Stage + PR status) and dependency edges (from each Phase's `depends_on`). Consumed
by the sessions/projects routers (T7/T8) and, ultimately, the frontend flow view.
"""
from __future__ import annotations

from typing import Optional

from backend.services.session_reader import EpicGroup, ParsedSession


def _pr_fields(session: ParsedSession) -> tuple[str, Optional[str]]:
    """Return (pr_status, pr_url). A present PR artifact means the PR is open."""
    pr = session.phases.get("pr", {}) if isinstance(session.phases, dict) else {}
    if not isinstance(pr, dict):
        pr = {}
    pr_url = pr.get("artifact")
    pr_status = "open" if pr_url else pr.get("status", "pending")
    return pr_status, pr_url


def _node(phase_index: int, phase_label: Optional[str], session: ParsedSession) -> dict:
    pr_status, pr_url = _pr_fields(session)
    return {
        "phase_index": phase_index,
        "phase_label": phase_label,
        "current_stage": session.current_phase,
        "pr_status": pr_status,
        "pr_url": pr_url,
    }


def build_epic_flow(group: EpicGroup) -> dict:
    """Build {epic, nodes, edges} for one Epic group.

    - nodes: one per child Phase (ordered by phase_index). An Epic with no child
      Phases (legacy/single-Phase, D9) yields a single synthetic node at index 1.
    - edges: {"from": d, "to": idx} for every d in each Phase's depends_on.
    - epic: {slug, label} or None for an orphan group (parent Epic missing).
    """
    epic_block = (
        {"slug": group.epic.slug, "label": group.epic.idea or group.epic.slug}
        if group.epic is not None
        else None
    )

    if not group.phases:
        # Legacy / single-Phase Epic: synthesize one node from the Epic itself.
        if group.epic is None:
            return {"epic": None, "nodes": [], "edges": []}
        node = _node(1, group.epic.idea or group.epic.slug, group.epic)
        return {"epic": epic_block, "nodes": [node], "edges": []}

    nodes: list[dict] = []
    edges: list[dict] = []
    for phase in group.phases:
        idx = phase.phase_index if phase.phase_index is not None else 0
        nodes.append(_node(idx, phase.phase_label, phase))
        for dep in phase.depends_on:
            edges.append({"from": dep, "to": idx})

    return {"epic": epic_block, "nodes": nodes, "edges": edges}

File: dashboard/frontend/src/lib/epicFlow.ts
// Client-side Epic grouping + flow derivation, mirroring backend
// services/epic_flow.py so the flow view can render from the flat session list
// (no extra fetch). See dashboard-epic-phase-view design (Phase B).
import type { Session, EpicFlow, PhaseNode } from "../types";

export interface EpicGroup {
  epic: Session | null;
  phases: Session[];
}

export function groupByEpic(sessions: Session[]): Record<string, EpicGroup> {
  const groups: Record<string, EpicGroup> = {};

  // Register every Epic (non-phase) first so phases can find their parent.
  for (const s of sessions) {
    if (s.kind !== "phase") {
      groups[s.slug] ??= { epic: s, phases: [] };
    }
  }
  // Attach phases to their parent (or an orphan group keyed by parentSlug).
  for (const s of sessions) {
    if (s.kind !== "phase") continue;
    const key = s.parentSlug ?? s.slug;
    (groups[key] ??= { epic: null, phases: [] }).phases.push(s);
  }
  for (const g of Object.values(groups)) {
    g.phases.sort((a, b) => (a.phaseIndex ?? Infinity) - (b.phaseIndex ?? Infinity));
  }
  return groups;
}

function prFields(s: Session): { pr_status: string; pr_url: string | null } {
  const pr = s.phases?.pr ?? {};
  const pr_url = (pr as any).artifact ?? null;
  return { pr_status: pr_url ? "open" : ((pr as any).status ?? "pending"), pr_url };
}

function node(phaseIndex: number, label: string | null, s: Session): PhaseNode {
  return {
    phase_index: phaseIndex,
    phase_label: label,
    current_stage: s.currentStage,
    ...prFields(s),
  };
}

export function toEpicFlow(group: EpicGroup): EpicFlow {
  const epic = group.epic
    ? { slug: group.epic.slug, label: group.epic.idea || group.epic.slug }
    : null;

  if (group.phases.length === 0) {
    if (!group.epic) return { epic: null, nodes: [], edges: [] };
    return { epic, nodes: [node(1, group.epic.idea || group.epic.slug, group.epic)], edges: [] };
  }

  const nodes: PhaseNode[] = [];
  const edges: { from: number; to: number }[] = [];
  for (const p of group.phases) {
    const idx = p.phaseIndex ?? 0;
    nodes.push(node(idx, p.phaseLabel ?? null, p));
    for (const dep of p.dependsOn) edges.push({ from: dep, to: idx });
  }
  return { epic, nodes, edges };
}
